lib/credo_core_node/mining/coinbase_transaction.ex

```
@@ -0,0 +1,136 @@
+        to: Mining.my_miner().address,
+        value: Pool.sum_pending_transaction_fees(txs),
+        fee: 1.0,
+        data: data
+      })
+
+    txs ++ [tx]
+  end
+
+  defp coinbase_tx_data(%{number: 0}), do: %{tx_type: Blockchain.coinbase_tx_type()}
+
+  defp coinbase_tx_data(previous_block) do
+    participation_rates =
+      previous_block
+      |> VotingCoordinator.calculate_participation_rates()
+      |> Enum.reject(fn {_address, participation_rate} ->
+        participation_rate >= VotingCoordinator.max_participation_rate()
+      end)
+      |> Map.new()
+
+    previous_votes_hash = Mining.get_votes_hash(previous_block, :hex)
+
+    %{
+      tx_type: Blockchain.coinbase_tx_type(),
+      participation_rates: participation_rates,
+      previous_votes_hash: previous_votes_hash
+    }
+  end
+
+  def valid_coinbase_transaction?(%{number: number}, _block_type) when number == 0, do: true
+
+  def valid_coinbase_transaction?(block, block_type) do
+    coinbase_txs = get_coinbase_txs(block)
+
+    res =
+      valid_number_of_txs?(coinbase_txs) && valid_value?(block, coinbase_txs) &&
+        valid_to?(coinbase_txs)
+
+    votes =
+      (block.number - 1)
+      |> Blockchain.get_block_by_number()
+      |> Mining.list_votes()
+
+    if Enum.any?(votes) && block_type == :pending do
+      res && valid_previous_votes_hash?(block, coinbase_txs) &&
+        valid_participation_rates?(block)
+    else
+      res
+    end
+  end
+
+  def valid_number_of_txs?(coinbase_txs), do: length(coinbase_txs) == 1
+
+  def valid_value?(block, coinbase_txs), do: tx_fee_sums_match?(block, coinbase_txs)
+
+  def valid_to?(coinbase_txs) do
+    to = List.first(coinbase_txs).to
+
+    !is_nil(Mining.get_miner(to))
+  end
+
+  def valid_previous_votes_hash?(%{number: 1} = _blk, _coinbase_txs), do: true
+
+  def valid_previous_votes_hash?(%{number: number} = _blk, coinbase_txs) do
+    previous_votes_hash =
+      (number - 1)
+      |> Blockchain.get_block_by_number()
```

Here we assume that `Blockchain.get_block_by_number/1` always returns a block but if could be false in case `number` is incorrect or the node doesn't have previous block for any reason (e.g. syncing with the network). How about adding a check that the block exists and returning false in case it does not?

---

@i1t I think this check is unnecessary:
1) current block won't pass validation on syncing if a previous block doesn't exist
2) if for whatever reason we pass invalid number, it would be better for function to crash rather than silently return `false`